File: Python/day14/AOC.py
import re
import time

def part1(input, width=101, height=103):
    grid = [[0 for _ in range(width)] for _ in range(height)]
    robots = [[x, y, vx, vy] for x, y, vx, vy in [map(int, re.findall(r"(-?\d+)", line)) for line in input]]
    qs = [0,0,0,0]
    w_half = width // 2
    h_half = height // 2
    for r in robots:
        r[0] += 100*r[2]
        r[0] %= width
        r[1] += 100*r[3]
        r[1] %= height
        grid[r[1]][r[0]] += 1

        if r[0] < w_half and r[1] < h_half:
            qs[0] += 1
        elif r[0] > w_half and r[1] < h_half:
            qs[1] += 1
        elif r[0] < w_half and r[1] > h_half:
            qs[2] += 1
        elif r[0] > w_half and r[1] > h_half:
            qs[3] += 1
    return qs[0] * qs[1] * qs[2] * qs[3]

# Pervious brute force solution showed me that the easter egg is within a box of robots, 
# so just find part of the box to find when the second the easter egg is revealed.
def part2(input):
    width = 101
    height = 103
    robots = [[x, y, vx, vy] for x, y, vx, vy in [map(int, re.findall(r"(-?\d+)", line)) for line in input]]

    for i in range(100000):
        grid = [['.' for _ in range(width)] for _ in range(height)]
        for r in robots:
            grid[r[1]][r[0]] = '#'
            r[0] += r[2]
            r[0] %= width
            r[1] += r[3]
            r[1] %= height

        for h in range(height):
            if re.search(r'\#{10,}', ''.join(grid[h])): # 10 consecutive '#', because that sus
                for row in grid:
                    print(''.join(row))
                return i

    return 0

def main():
    input = open("input.txt", "r").read().splitlines()
    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

if __name__ == "__main__":
    main()

File: Go/day14/AOC.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

// Helper function to calculate the modulus of a negative number
func mod(a, b int) int {
	return (((a % b) + b) % b)
}

type Point struct {
	x, y int
}

type Robot struct {
	x, y   int
	vx, vy int
}

func (r *Robot) move(times, width, height int) {
	r.x = mod(r.x+times*r.vx, width)
	r.y = mod(r.y+times*r.vy, height)
}

func parseInput(input []string) []Robot {
	Robots := []Robot{}
	for _, line := range input {
		matches := regexp.MustCompile(`(-?\d+)`).FindAllString(line, -1)
		nums := make([]int, len(matches))
		for i, match := range matches {
			fmt.Sscanf(match, "%d", &nums[i])
		}
		x, y, vx, vy := nums[0], nums[1], nums[2], nums[3]
		Robots = append(Robots, Robot{x, y, vx, vy})
	}
	return Robots
}

func part1(input []string, test bool) int {
	Width, Height := 101, 103
	if test {
		Width, Height = 11, 7
	}
	Robots := parseInput(input)
	qudrants := make([]int, 4)

	halfWidth, halfHeight := Width/2, Height/2

	for _, r := range Robots {
		r.move(100, Width, Height)
		if r.x == halfWidth || r.y == halfHeight {
			continue
		}
		if r.x < halfWidth {
			if r.y < halfHeight {
				qudrants[0]++
			} else {
				qudrants[1]++
			}
		} else {
			if r.y < halfHeight {
				qudrants[2]++
			} else {
				qudrants[3]++
			}
		}
	}
	return qudrants[0] * qudrants[1] * qudrants[2] * qudrants[3]
}

// There is a way to solve this by looking at the safety factor of the robots,
// but that approach does not seem to work for my input. So I just brute force.
func part2(input []string) int {
	Width, Height := 101, 103
	Robots := parseInput(input)
	t := 0
	for true {
		t++
		pos := make(map[Point]bool)
		valid := true
		for _, r := range Robots {
			x := mod(r.x+t*r.vx, Width)
			y := mod(r.y+t*r.vy, Height)
			if pos[Point{x, y}] {
				valid = false
				break
			}
			pos[Point{x, y}] = true
		}
		if valid {
			break
		}
	}
	return t
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines, false)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
